Scheduler fixes and improvements

mateiz opened the pull request:
This is a work in progress to simplify and improve a few things in the TaskSet scheduler and the way we handle locality. Here's a summary of changes (see also the commit messages):
- Eliminated most hostPort based data structures and just used executor IDs instead, since the hostPort vs host stuff was confusing (and not checkable with static typing, leading to ugly debug code), and hostPorts are not provided by Mesos. The hostPort stuff was there from before when we could easily pass executorIDs through.
- Removed non-local fallback stuff in ClusterScheduler that tried to launch less-local tasks on a node once the local ones were all assigned. This change didn't work because many cluster schedulers send offers for just one node at a time (even the standalone and YARN ones do so as nodes join the cluster one by one). Thus, lots of non-local tasks would be assigned even though a node with locality for them would be able to receive tasks just a short time later.
- Added multi-level delay scheduling in ClusterTaskSetManager so you can have different waits for each locality level. This covers part of the use case for the fallback stuff above (for users who wanted to skip node-local waits and just go for rack-local).
- Updated the way ClusterTaskSetManager handles racks: instead of enqueueing a task to a separate queue for all the hosts in the rack, which would create lots of large queues, have one queue per rack name.
- Added periodic revival of offers in StandaloneSchedulerBackend (this was an option in the YARN patch but seems good to have by default).
- Renamed map output tracker "generations" to "epochs".
- Simplified and cleaned up a bunch of the code.

The only thing missing is new unit tests for this stuff, which I'm working on, but the code appears to run fine so I wanted to put it up for others to try.

mateiz committed:
Initial work towards scheduler refactoring:

- Replace use of hostPort vs host in Task.preferredLocations with a
  TaskLocation class that contains either an executorId and a host or
  just a host. This is part of a bigger effort to eliminate hostPort
  based data structures and just use executorID, since the hostPort vs
  host stuff is confusing (and not checkable with static typing, leading
  to ugly debug code), and hostPorts are not provided by Mesos.

- Replaced most hostPort-based data structures and fields as above.

- Simplified ClusterTaskSetManager to deal with preferred locations in a
  more concise way and generally be more concise.

- Updated the way ClusterTaskSetManager handles racks: instead of
  enqueueing a task to a separate queue for all the hosts in the rack,
  which would create lots of large queues, have one queue per rack name.

- Removed non-local fallback stuff in ClusterScheduler that tried to
  launch less-local tasks on a node once the local ones were all
  assigned. This change didn't work because many cluster schedulers send
  offers for just one node at a time (even the standalone and YARN ones
  do so as nodes join the cluster one by one). Thus, lots of non-local
  tasks would be assigned even though a node with locality for them
  would be able to receive tasks just a short time later.

- Renamed MapOutputTracker "generations" to "epochs".

mateiz committed:
More scheduling fixes:

- Added periodic revival of offers in StandaloneSchedulerBackend

- Replaced task scheduling aggression with multi-level delay scheduling
  in ClusterTaskSetManager

- Fixed ZippedRDD preferred locations because they can't currently be
  process-local

- Fixed some uses of hostPort

mateiz committed:
Comment cleanup (via Kay) and some debug messages

AmplabJenkins commented:
Thank you for submitting this pull request.

All automated tests for this request have passed.

Refer to this link for build results: http://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/581/

pwendell commented on core/src/main/scala/spark/scheduler/cluster/ClusterTaskSetManager.scala:
```diff
@@ -801,4 +655,39 @@ private[spark] class ClusterTaskSetManager(sched: ClusterScheduler, val taskSet:
   }
+
+  private def getLocalityWait(level: TaskLocality.TaskLocality): Long = {
+    val defaultWait = System.getProperty("spark.locality.wait", "3000")
```
Matei - what are the semantics of these different waits? Say that you want a task to bump from PROCESS_LOCAL to NODE_LOCAL after 10 seconds. Would you set the wait to 10 seconds for PROCESS_LOCAL or NODE_LOCAL? It would be good to make that clear in the docs.

It might make sense to have the default wait be different for different levels so that the "turn key" performance is pretty good. For instance, it's probably fine to make the PROCESS_LOCAL -> NODE_LOCAL switch after only a few seconds. But going from PROCESS_LOCAL to RACK_LOCAL should be a lot longer.

mateiz replied:
Good idea; I'll add a doc page on this. Basically there is a separate jump between each pair of levels: for example process -> (wait 3 sec) -> node -> (wait 3 more sec) -> rack. You can configure each wait separately.

My instinct was that it's not obvious which jumps will cost more, so I'd rather have a simple default setting. For example, going from process to node could be really bad for data that's expensive to serialize, while going from node to rack might be very cheap with a fast network.

pwendell replied:
I see - so I guess my main question then was whether this defines the level we are jumping _to_ or jumping _from_. I guess since you can't set this for `ANY` it's the level you are jumping _from_.

mridulm commented on core/src/main/scala/spark/deploy/worker/ExecutorRunner.scala:
```diff
@@ -40,13 +40,11 @@ private[spark] class ExecutorRunner(
     val workerId: String,
-    val hostPort: String,
+    val host: String,
     val sparkHome: File,
```
I am not very sure about this - iirc this needs to be hostport - since we can have multiple executors in the same node.

mateiz replied:
It's only used to set the host below in this file (no other references), so it should be fine.

mridulm replied:
can we remove references to host also ?
iirc we can infer the host from the executorId right ? This will remove all references to host when we can avoid it ...

mridulm commented on core/src/main/scala/spark/rdd/ZippedPartitionsRDD.scala:
```diff
@@ -55,27 +55,15 @@ abstract class ZippedPartitionsBaseRDD[V: ClassManifest](
+    } else {
+      prefs.flatten.distinct
+    }
   }
```
If there are both "host:port" and "host" preferred locations, this will result in loosing the "host" preferred locations - right ?
Will that happen - I am not sure, but we should be defending against future changes imo.

Same applies to other similar changes ...

mateiz replied:
No, I've decided not to allow host:port in preferred locations. Instead, Tasks that have a preferred executor can pass an executorID as part of their TaskLocation object.

mridulm replied:
ah, that is an excellent change ! had not noticed that

mridulm commented on core/src/main/scala/spark/scheduler/DAGSchedulerEvent.scala:
```diff
@@ -54,9 +54,7 @@ private[spark] case class CompletionEvent(
-  Utils.checkHostPort(hostPort, "Required hostport")
-}
+private[spark] case class ExecutorGained(execId: String, host: String) extends DAGSchedulerEvent
 
```
How will this handle the case of multiple containers in a single node ? (same host, different ports)

mateiz replied:
Using the executorID. That's how it gets used below.

mridulm commented on core/src/main/scala/spark/scheduler/cluster/ClusterScheduler.scala:
```diff
@@ -169,27 +127,6 @@ private[spark] class ClusterScheduler(val sc: SparkContext)
-        }
-      }.start()
-    }
   }
```
The rationale behind this thread was to revive offers periodically : and not necessarily have to wait for a task event (finished/failed/etc) to reissue revive requests for all 'free' workers.

mateiz replied:
I've moved this into StandaloneSchedulerBackend. Specifically, that has a new line to schedule a ReviveOffers message to itself every 1000 ms (or whatever the timeout is)

mridulm replied:
cool ! i missed out on this, my apologies.

mridulm commented on core/src/main/scala/spark/scheduler/cluster/ClusterScheduler.scala:
```diff
@@ -235,172 +172,53 @@ private[spark] class ClusterScheduler(val sc: SparkContext)
+      if (!executorsByHost.contains(o.host)) {
+        executorsByHost(o.host) = new HashSet[String]()
+        executorGained(o.executorId, o.host)
       }
```
resourceOffers can be invoked from different threads (akka and others too ?), so we might want to have this around sync block - not just for locking, but also to ensure changes are reflected across threads.

mateiz replied:
Yeah, sorry it's not super clear, but I just changed the code style to add it at the top:

```
def resourceOffers(offers: Seq[WorkerOffer]): Seq[Seq[TaskDescription]] = synchronized { ...
```

This is a more common style for making a whole method synchronized in Scala

mridulm replied:
There are changes in ClusterTaskSetManager also which removed synch ... I think I added the comment at the wrong location.
Ideally ClusterTaskSetManager should only get invoked from here, but they were added to be defensive in case of future changes for consistency of update to shared state.

... and I had not noticed the synchronized at the top of method definition - thanks for pointing that out btw !

mridulm commented on core/src/main/scala/spark/scheduler/cluster/ClusterScheduler.scala:
```diff
@@ -536,80 +353,56 @@ private[spark] class ClusterScheduler(val sc: SparkContext)
 
     // order keyList based on population of value in map
     val keyList = _keyList.sortWith(
+      // TODO(matei): not sure why we're using getOrElse if keyList = map.keys... see if it matters
```
getOrElse is not really required - was probably an oversight on my part.
I think it was one of those templatized copy/paste from java to scala code conversions.

mridulm commented on core/src/main/scala/spark/scheduler/cluster/ClusterTaskSetManager.scala:
```diff
@@ -171,166 +127,86 @@ private[spark] class ClusterTaskSetManager(sched: ClusterScheduler, val taskSet:
-          }
+  // Figure out which locality levels we have in our TaskSet, so we can do delay scheduling
+  val myLocalityLevels = computeValidLocalityLevels()
+  val localityWaits = myLocalityLevels.map(getLocalityWait) // Time to wait at each level
```
Unless I am misreading the diffs - computeValidLocalityLevels, etc is a val in the class right ?
Wont that not need to be recomputed each time tasks are added/finished ?

mateiz replied:
Yeah, for simplicity I decided not to recompute it. Technically we could optimize things a bit by jumping straight to the next locality level when we exhaust stuff at one level (for example, if we had some process-local tasks and some that could only be node-local, we should jump to the latter as soon as we're done scheduling the process-local ones). I don't think the number of levels ever needs to increase, since even if nodes fail, tasks will only be added to a subset of the previous levels or the noPrefs list (e.g. a task that had both process- and node-local locations might now only have node). I'll look into this.

mridulm commented on core/src/main/scala/spark/scheduler/cluster/ClusterTaskSetManager.scala:
```diff
@@ -171,166 +127,86 @@ private[spark] class ClusterTaskSetManager(sched: ClusterScheduler, val taskSet:
+          addTo(pendingTasksForRack.getOrElseUpdate(rack, new ArrayBuffer))
+        }
+        hadAliveLocations = true
       }
```
Will this handle the case of new workers coming up on the same host or rack (for this block and the block above) ?
I was not able to come to a conclusion either way ...

mateiz replied:
Because we're creating queues based on the host name and rack name now, this should work, as long as sched.getRackForHost returned the right rack name.

mridulm replied:
Looks good, thanks for clarifying !

mridulm commented on core/src/main/scala/spark/scheduler/cluster/ClusterScheduler.scala:
```diff
@@ -235,172 +172,53 @@ private[spark] class ClusterScheduler(val sc: SparkContext)
+          val execId = offers(i).executorId
+          val host = offers(i).host
+          for (task <- manager.resourceOffer(execId, host, availableCpus(i))) {
+            tasks(i) += task
```
This actually leads to potentially suboptimal scheduling.
Consider the case of 4 workers ready for scheduling, with the level set to rack local (anything worse than instance local).
The best scheduling will be in the order instance local, node local, rack local, others.

But that need not necessarily be gauranteed here - for example, if task1 has offers(0) rack local, offers(1) node local and offers(2) instance local.
Ideally, we will want task1 to be scheduled on offers(2), not even on offers(1) ... which is what the earlier messy blocks were attempting to do - to minimize intra node (added as an after thought thanks to @rxin's suggestion !), followed by inter node, and then intra-rack data movement.

mateiz replied:
This is a good point, though it will only happen if some TaskSets have already waited enough to go to a non-local level and a local node happens to become available in the same set of resource offers. It can't really happen when the cluster is busy and only one node is becoming free at a time  (the most common case), or when the cluster was idle and a new job is starting (because then it won't have waited). That said though, I'll see if I can change this to loop through the locality levels as before. My main concern was that I didn't want to launch _non-local_ tasks as soon as we exhaust the local spots in this resource offer if those TaskSets haven't waited enough.

mridulm replied:
When we have individual offers being made, you are right - this will not be an issue : since we have only a single executor being offered anyway.
But at start of a job, when we have a bunch of offers, this will result in suboptimal scheduling (primarily between instance local and node local) - but can also happen when multiple nodes are passed on to resource offers - when we have node failures and re-allocations, etc.

mridulm replied:
To add to above (late night comments are not always very optimal :) )
There were two intentions behind the earlier piece of code :

a) Given a seq of WorkerOffer, ensure that the order in which they are passed on to TaskSetManager is such that relatively optimal scheduling of workers occurs.
Namely, prioritize instance local, node local, rack local and any in that order.

b) Once all optimal schedules have happened, expand the scope so that non optimal schedules are also done such that cluster utilization is mazimized (within bounds controlled via aggression and time spent waiting).
These are also done in order of our prioritity - so even if aggression was set to ANY, we will try to do rack local schedules first, before ANY.

This tries to ensure that we balance optimal scheduling with maximizing utilization of cluster resources.
Invocation of resourceOffers with single workoffer happens when a task finishes/fails; which is unfortunately not very optimal - I never got around to resolving this and relying only on reviveOffers() (bulk submission of all available workers)

Regards,
Mridul

mateiz replied:
Yes, I'll implement point a), but for point b), as I said above, it's not a good idea. Single worker offers are the common case while a job is running (e.g. as it finishes the first series of tasks and launches another), and it fundamentally wouldn't make sense to delay them to wait for multiple workers to finish because then you'd be wasting cluster resources. It's not just an issue when a job begins -- this aggression logic degrades locality throughout the job runtime. For example, if you have a job with cached data (process-local) that is running multiple waves of tasks, some tasks will be scheduled non-locally in the current code, leading to a very expensive transfer of a block across nodes when you could've just waited a few seconds to run on the right node. With my new code, it's still very easy to minimize the initial wait if that's a concern, and you get the additional bonus of making the process-local wait time different from the node/rack one.

mridulm replied:
Hmm, on the contrary, the aggression logic is supposed to make it user
controllable on whether the tradeoff of waiting for optimal scheduling vs
maximising cluster utilization while maximising locality on the set of one
or more offers.
Particularly with locality wait and larger number of workers, the
reviveOffers thread will be invoking the method with nontrivial number of
offers - optimizing the single offer case was left for later (see below).

In our tests (ml apps with disk or mem ser on non trivial input datasets)
introduction of this one change speedup of jobs by 2.5x - 4x (iirc) with no
other change in system : with setting it to the default having no perf
change w.r.t earlier perf of jobs in our measurements.
As I mentioned before, idling large number of workers is extremely
suboptimal for our usecases.
I am sure for a large number of other applications of spark this is not a
concern : if entire working set fits in memory and is hosted in non-ser
mode, etc - which is why it would be good to allow its customization via
flags - while keeping it at defaults to handle common case.

Re single worker case : we can simply disable aggression code for this
(earlier, the last iteration expanded the locality constraint) when in this
last iteration and num offers = 1 : this was not done since reviveOffers
thread was optional unlike now.
This should disallow non host or instance local scheduling until timeout
for single offer case.

Regards
Mridul
 On Aug 16, 2013 3:33 AM, "Matei Zaharia" notifications@github.com wrote:

> In core/src/main/scala/spark/scheduler/cluster/ClusterScheduler.scala:
> 
> > ```
> >      for (i <- 0 until offers.size) {
> > ```
> > -          val hostPort = offers(i).hostPort
> > -          // DEBUG code
> > 
> > ## \-          Utils.checkHostPort(hostPort)
> > -          val numProcessLocalTasks =  math.max(0, math.min(manager.numPendingTasksForHostPort(hostPort), availableCpus(i)))
> > -          if (numProcessLocalTasks > 0){
> > -            val list = processLocalOffers.getOrElseUpdate(hostPort, new ArrayBuffer[Int])
> > -            for (j <- 0 until numProcessLocalTasks) list += i
> > -          val execId = offers(i).executorId
> > -          val host = offers(i).host
> > -          for (task <- manager.resourceOffer(execId, host, availableCpus(i))) {
> > -            tasks(i) += task
> 
> Yes, I'll implement point a), but for point b), as I said above, it's not
> a good idea. Single worker offers are the common case while a job is
> running (e.g. as it finishes the first series of tasks and launches
> another), and it fundamentally wouldn't make sense to delay them to wait
> for multiple workers to finish because then you'd be wasting cluster
> resources. It's not just an issue when a job begins -- this aggression
> logic degrades locality throughout the job runtime. For example, if you
> have a job with cached data (process-local) that is running multiple tasks
> per core, some tasks will be scheduled non-locally in the current code,
> leading to a very expensive transfer of a block across nodes when you
> could've just waited a few of seconds to run on the right node. With my new
> code, it's still very easy to minimize the initial wait if that's a
> concern, and you get the additional bonus of making the process-local wait
> time different from the node/rack one.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/mesos/spark/pull/828/files#r5802888
> .

mateiz replied:
As I said, you can get the same behavior by controlling the different timeouts. As far as I understood it, the biggest change for you was actually from skipping node locality to go to rack; this can be done easily by setting the node locality wait to 0. If you also want to skip process locality, you can try that too, though I don't think this is usually a good idea. The new code is also simpler than the old aggression code, more efficient (doesn't allocate a whole bunch of hashmaps on every resource offer), and avoids some real performance regressions I saw with that change. If you guys try it out and see some performance problems that can't be solved with these settings, I'm happy to revisit it.

BTW, I've also pushed an update for this that now goes through the locality levels in order for the whole set of available offers.

mridulm replied:
Let me clarify the performance issue we face, in the current code, if we
set node locality wait to 0 - the scaling is extremely bad due to heavy
data movement at job start (it is worse than non zero node locality value
actually - so it is pointless to set that atleast in our expts).
Particularly when dataset is large. Effectively, almost all schedules end
up being rack local.

Setting it to anything non zero (however small - in our expts even 1ms -
since effectively it is a much higher wait), the cluster utilization drops
drastically.
Before introducing this change, we had experimented with reducing locality
wait to 0ms and 1ms in earlier codebase - and since that did not help
matters, we modified the scheduling ... initially aggression was
effectively any - which improved situation much better, and adding rack
support improved it even further while minimizing io overhead but at
slightly decreased cluster utilization (so it was a tradeoff between both).

Scheduling is something we need to heavily experiment with - considering
the non trivial gains we saw with relatively small changes.

On Sat, Aug 17, 2013 at 1:47 PM, Matei Zaharia notifications@github.comwrote:

> In core/src/main/scala/spark/scheduler/cluster/ClusterScheduler.scala:
> 
> > ```
> >      for (i <- 0 until offers.size) {
> > ```
> > -          val hostPort = offers(i).hostPort
> > -          // DEBUG code
> > 
> > ## \-          Utils.checkHostPort(hostPort)
> > -          val numProcessLocalTasks =  math.max(0, math.min(manager.numPendingTasksForHostPort(hostPort), availableCpus(i)))
> > -          if (numProcessLocalTasks > 0){
> > -            val list = processLocalOffers.getOrElseUpdate(hostPort, new ArrayBuffer[Int])
> > -            for (j <- 0 until numProcessLocalTasks) list += i
> > -          val execId = offers(i).executorId
> > -          val host = offers(i).host
> > -          for (task <- manager.resourceOffer(execId, host, availableCpus(i))) {
> > -            tasks(i) += task
> 
> As I said, you can get the same behavior by controlling the different
> timeouts. As far as I understood it, the biggest change for you was
> actually from skipping node locality to go to rack; this can be done easily
> by setting the node locality wait to 0. If you also want to skip process
> locality, you can try that too, though I don't think this is usually a good
> idea. The new code is also simpler than the old aggression code, more
> efficient (doesn't allocate a whole bunch of hashmaps on every resource
> offer), and avoids some real performance regressions I saw with that
> change. If you guys try it out and see some performance problems that can't
> be solved with these settings, I'm happy to revisit it.
> 
> BTW, I've also pushed an update for this that now goes through the
> locality levels in order for the whole set of available offers.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/mesos/spark/pull/828/files#r5828582
> .

mateiz replied:
Mridul, as I said, please try it with my change and spark.locality.wait.node = 0. If you want the exact same behavior as your old code, you can also set spark.locality.wait.process=0. The behavior will be _exactly_ the same:
- When multiple offers come in, Spark will look at _all_ of them together and try to assign only process-local tasks first, then node-local, then rack-local, etc. Note that I added the logic to go through these levels in order across the set of offers.
- If you set both node and process waits to 0, it will immediately jump up to rack to fill in all the remaining spots.

I really don't see how it's different, except that I'm doing this with significantly simpler code.

I've been working on similar stuff in Hadoop for a long time (wrote the Hadoop Fair Scheduler in 2008, which uses the same algorithm for locality and has been running fine on 1000+ node clusters since then at Facebook, many Cloudera users, etc), so it's not like this is an untested change. There was definitely a problem before for large clusters where you don't have executors on most of the data-containing nodes, but the combination of rack-awareness and the ability to configure these delays and to optimize across a set of offers fixes that. I'm not trying to remove code because I don't understand it -- I just think that it's important for this logic (and all of Spark's scheduling code) to be dead-simple in order to be easy to optimize and understandable to new developers.

> Let me clarify the performance issue we face, in the current code, if we set node locality wait to 0 - the scaling is extremely bad due to heavy data movement at job start (it is worse than non zero node locality value actually - so it is pointless to set that atleast in our expts). Particularly when dataset is large. Effectively, almost all schedules end up being rack local. Setting it to anything non zero (however small - in our expts even 1ms - since effectively it is a much higher wait), the cluster utilization drops drastically. Before introducing this change, we had experimented with reducing locality wait to 0ms and 1ms in earlier codebase - and since that did not help matters, we modified the scheduling ... initially aggression was effectively any - which improved situation much better, and adding rack support improved it even further while minimizing io overhead but at slightly decreased cluster utilization (so it was a tradeoff between both). Scheduling is something we need to heavily experiment with - considering the non trivial gains we saw with relatively small changes.

mridulm replied:
Ah, I see my disconnect now.
The local PR I had was not updated with commit 1d6da02 - hence the bad perf
numbers I observed (it takes a couple of days to finish) : and commented
based on analysis of the local code - my apologies !
Preliminary results seem to indicate reasonably close enough numbers -
thanks for clarifying.

On Sat, Aug 17, 2013 at 10:23 PM, Matei Zaharia notifications@github.comwrote:

> In core/src/main/scala/spark/scheduler/cluster/ClusterScheduler.scala:
> 
> > ```
> >      for (i <- 0 until offers.size) {
> > ```
> > -          val hostPort = offers(i).hostPort
> > -          // DEBUG code
> > 
> > ## \-          Utils.checkHostPort(hostPort)
> > -          val numProcessLocalTasks =  math.max(0, math.min(manager.numPendingTasksForHostPort(hostPort), availableCpus(i)))
> > -          if (numProcessLocalTasks > 0){
> > -            val list = processLocalOffers.getOrElseUpdate(hostPort, new ArrayBuffer[Int])
> > -            for (j <- 0 until numProcessLocalTasks) list += i
> > -          val execId = offers(i).executorId
> > -          val host = offers(i).host
> > -          for (task <- manager.resourceOffer(execId, host, availableCpus(i))) {
> > -            tasks(i) += task
> 
> Mridul, as I said, please try it with my change and
> spark.locality.wait.node = 0. If you want the exact same behavior as your
> old code, you can also set spark.locality.wait.process=0. The behavior will
> be _exactly_ the same: - When multiple offers come in, Spark will look at
> _all_ of them together and try to assign only process-local tasks first,
> then node-local, then rack-local, etc. Note that I added the logic to go
> through these levels in order across the set of offers. - If you set both
> node and process waits to 0, it will immediately jump up to rack to fill in
> all the remaining spots. I really don't see how it's different, except that
> I'm doing this with significantly simpler code.
> … <#1408d31ed1554ae1_>
> I've been working on similar stuff in Hadoop for a long time (wrote the
> Hadoop Fair Scheduler in 2008, which uses the same algorithm for locality
> and has been running fine on 1000+ node clusters since then at Facebook,
> many Cloudera users, etc), so it's not like this is an untested change.
> There was definitely a problem before for large clusters where you don't
> have executors on most of the data-containing nodes, but the combination of
> rack-awareness and the ability to configure these delays and to optimize
> across a set of offers fixes that. I'm not trying to remove code because I
> don't understand it -- I just think that it's important for this logic (and
> all of Spark's scheduling code) to be dead-simple in order to be easy to
> optimize and understandable to new developers.
>  Let me clarify the performance issue we face, in the current code, if we
> set node locality wait to 0 - the scaling is extremely bad due to heavy
> data movement at job start (it is worse than non zero node locality value
> actually - so it is pointless to set that atleast in our expts).
> Particularly when dataset is large. Effectively, almost all schedules end
> up being rack local. Setting it to anything non zero (however small - in
> our expts even 1ms - since effectively it is a much higher wait), the
> cluster utilization drops drastically. Before introducing this change, we
> had experimented with reducing locality wait to 0ms and 1ms in earlier
> codebase - and since that did not help matters, we modified the scheduling
> ... initially aggression was effectively any - which improved situation
> much better, and adding rack support improved it even further while
> minimizing io overhead but at slightly decreased cluster utilization (so it
> was a tradeoff betwe en both) . Scheduling is something we need to heavily
> experiment with - considering the non trivial gains we saw with relatively
> small changes.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/mesos/spark/pull/828/files#r5829281
> .

mridulm commented:
Hi Matei,

  I have added a few comments, unfortunately I am unable to go through this at the level of detail I would like to - apologies for that.

In general, a few things though :

a) There were a few sync blocks removed - we might need to re-introduce those back.
(Not just for MT-safety, which might not be a big issue, but also to ensure visibility of updates since same thread need not necessarily be used by akka/etc).

b) I noticed change from hostPort to host, executorId - would it not be better to remove host also ?
The earlier code was more of a hack - since I was trying to reconcile pre-executorId codebase with significant drift in yarn branch which had host -> hostPort change.

c) I added a few comments about scheduling in general in the code, and why we hacked in reviveOffers thread : coupled with my earlier mail on this subject, hopefully I have not messed up explaining my rationale for those changes !

Thanks,
Mridul

mateiz commented:
Thanks for the feedback, Mridul, this is super helpful! I'll address some of the issues you brought up.

ryanlecompte commented on core/src/main/scala/spark/scheduler/cluster/ClusterTaskSetManager.scala:
```diff
@@ -171,166 +127,86 @@ private[spark] class ClusterTaskSetManager(sched: ClusterScheduler, val taskSet:
+   * Add a task to all the pending-task lists that it should be on. If readding is set, we are
+   * re-adding the task so only include it in each list if it's not already there.
+   */
+  private def addPendingTask(index: Int, readding: Boolean = false) {
```
minor typo: readding

ryanlecompte replied:
oh, nevermind! re-adding, not readding :)

mateiz committed:
Address some review comments:

- When a resourceOffers() call has multiple offers, force the TaskSets
  to consider them in increasing order of locality levels so that they
  get a chance to launch stuff locally across all offers

- Simplify ClusterScheduler.prioritizeContainers

- Add docs on the new configuration options

AmplabJenkins commented:
Thank you for submitting this pull request.

All automated tests for this request have passed.

Refer to this link for build results: http://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/610/

mateiz committed:
Added some comments on threading in scheduler code

AmplabJenkins commented:
Thank you for submitting this pull request.

All automated tests for this request have passed.

Refer to this link for build results: http://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/635/

mateiz committed:
Added unit tests for ClusterTaskSetManager, and fix a bug found with
resetting locality level after a non-local launch

mateiz commented:
Jenkins, please retest this

mateiz commented:
Alright, I think this is pretty much done; I also added some unit tests and found a bug with them (ClusterTaskSetManager was setting currentLocalityIndex to allowedLocality instead of taskLocality).

One final improvement might be to add support for rack locality in non-YARN modes, but we can do that as a separate PR.

AmplabJenkins commented:
Thank you for submitting this pull request.

All automated tests for this request have passed.

Refer to this link for build results: http://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/646/

AmplabJenkins commented:
Thank you for submitting this pull request.

All automated tests for this request have passed.

Refer to this link for build results: http://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/647/

mridulm commented:
I ran some tests with this PR (as pulled on saturday). The results were mixed - 

The setup:
a) baseline - spark.tasks.revive_periodic.interval=200, spark.tasks.schedule.aggression=RACK
b) pr applied, enabled - spark.locality.wait.process=0 spark.locality.wait.node=0 spark.locality.wait.rack=3000 (to mirror (a) as closely as possible).
c) pr applied, but not enabled - so run with default config.

The relative performance numbers below are for 4 runs of the expts, and both min and avg of the time taken via 'time' command (so end to end time).

For lower number of nodes ( <= 10) , (b) performed better : from 2% to 10% for the experiments. For (c), the results were break even to worse by 4% or so.

But as the number of nodes increase, the relative performance worsens, and somewhere between 20 to 25, (b) becomes worse than (a).
At 100 nodes, (b) is about 18 % worse and (c) is about 41% worse.

I did not try with large numbers though (partly due to diminishing returns for the dataset, and partly because the graphs were diverging anyway, so intutively it should worsen).

This is much better than earlier though (without 2a4ed10).
Also note that the relative difference between (b) and (c) is much lower than earlier (with and without aggression) - which can be taken as a positive change : default out of box performance is better.

I am planning to rerun the expt set once again with the latest PR, just to be sure.

mateiz commented:
I actually did fix a small bug in it yesterday, where it wasn't properly remembering the locality of each launched task.

Instead of c), can you also try setting process spark.locality.wait.process=3000, spark.locality.wait.node=0, and spark.locality.wait.rack=3000? This is a new configuration possible with this PR that lets you launch tasks on HDFS in a rack-local manner while still getting good locality for cached data. (Is your test actually using any caching by the way?)

mateiz commented:
Also, if you want to match a), you need to set spark.scheduler.revive.interval=200 in the new code too.

mridulm commented:
I was trying to replicate exactly similar test env for both baseline and the pr change (hence spark.locality.wait.process=0).

I did not notice that revive thread timeout was configurable - will change the test to use spark.scheduler.revive.interval = 200, thanks !

mridulm commented:
Looks great now Matei !
With the latest commits the preliminary numbers with PR enabled is within margin error with baseline.

mateiz commented:
Thanks for testing it, Mridul. It also does fine on Patrick's regression tests (https://github.com/amplab/spark-perf), so I'm going to merge it in.
